File: src/main/java/example/domain/model/specification/ExpressFare.java
package example.domain.model.specification;

import java.util.Objects;

import example.domain.model.bill.Amount;

/**
 * 特急料金
 */
public class ExpressFare {

    Destination destination;
    SeatType seatType;
    TrainType trainType;
    Amount fare;
    Amount additionalFare;

    public ExpressFare(Destination destination, SeatType seatType, Amount fare, Amount additionalFare) {
        this.destination = destination;
        this.seatType = seatType;
        this.fare = fare;
        this.additionalFare = additionalFare;
        this.trainType = TrainType.valueOf(additionalFare);
    }

    public Amount calculate() {
        return this.seatType.calculate(this.fare.add(additionalFare));
    }

    @Override
    public String toString() {
        return this.destination + "行き " + seatType.toString() + " " + trainType.toString() + " 料金" + fare.toString() + "";
    }

    @Override
    public boolean equals(Object other) {
        return isEqual((ExpressFare) other);
    }

    private boolean isEqual(ExpressFare otherFare) {
        return this.destination == otherFare.destination && this.seatType == otherFare.seatType && this.fare.equals(otherFare.fare);
    }

    @Override
    public int hashCode() {
        return Objects.hash(destination) + Objects.hash(seatType);
    }
}

File: jr-pricing/Model/ExpressFare.vb
Imports Jr.Pricing.Fw
Imports Jr.Pricing.Model.Bill
Imports Jr.Pricing.Model.Specification

Namespace Model
    ''' <summary>
    ''' 特急料金
    ''' </summary>
    ''' <remarks></remarks>
    Public Class ExpressFare : Inherits ValueObject

        Private ReadOnly destination As Destination
        Private ReadOnly seatType As SeatType
        Private ReadOnly trainType As TrainType
        Private ReadOnly fare As Amount
        Private ReadOnly additionalFare As Amount

        Public Sub New(destination As Destination, seatType As SeatType, fare As Amount, additionalFare As Amount)
            Me.destination = destination
            Me.seatType = seatType
            Me.fare = fare
            Me.additionalFare = additionalFare
            Me.trainType = trainType.ValueOf(additionalFare)
        End Sub

        Protected Overrides Function GetAtomicValues() As IEnumerable(Of Object)
            Return New Object() {destination, seatType, trainType, fare, additionalFare}
        End Function

        Public Function Calculate() As Amount
            Return Me.seatType.Calculate(Me.fare.Add(additionalFare))
        End Function

    End Class
End Namespace